[DSET-4037] [DSET-4014] [DSET-4034] [DSET-4034] [DSET-4098] Dataset log exporter improvements

tomaz-s1 opened the pull request:
This pull request contains various small improvements, changes and fixes for the DataSet Log exporter.

I bundled all the changes together since they are all on the small side and part of the general exporter clean up and improvements work.

### 1. Remove ``OtelExporter - Log -`` prefix from all the DataSet event ``message`` fields - DSET-4014
  
This prefix is not needed nor desired.

## 2. Correctly handle log record severity / log level - DSET-4037

I updated the code to correctly handle and map from OTel LogRecord severity to DataSet event severity (logic is described in the code comments).

I also removed now redundant and unnecessary event ``severity.number`` and ``severity.text`` attributes from the event.

## 3. Remove redundant flags and flags.is_sampled attribute - DSET-4034

Remove unnecessary and redundant `flags` and `flags.is_sampled` attribute.

## 4. Update metadata file to include `distributions` field - DSET-4098

Title is self explanatory.

tomaz-s1 committed:
Remove "OtelExporter - Log -" suffix from the event "message" field.
This prefix is nor desired nor wanted.

tomaz-s1 committed:
Implement severity field handling and make sure we currently map OTel
severity to DataSet severity.

Also remove "severity.text" and "severity.number" field from the event
since it's redundant - we already have event severity (sev) field value.

tomaz-s1 committed:
Remove flags and flags.is_sampled field since it provides no additional
information which is not already available (events which are sampled
will already be ingested and visible in the ui so flags.is_sampled is
redundant).

zdaratom-s1 commented on exporter/datasetexporter/logs_exporter.go:
```diff
@@ -65,6 +66,73 @@ func buildBody(attrs map[string]interface{}, value pcommon.Value) string {
 }
 
+// Function maps OTel severity on the LogRecord to DataSet severity level (number)
+func otelSeverityToDataSetSeverity(log plog.LogRecord) int {
```
what about mapOtelSeverityToDataSetSeverity ?

tomaz-s1 replied:
That's fine with me. 

I'm not an expert on go and otel conventions, but I do know they try to avoid "get" prefix in the function names so I left it out. I guess "map" may be fine.

tomaz-s1 replied:
From their docs:

> Go doesn't provide automatic support for getters and setters. There's nothing wrong with providing getters and setters yourself, and it's often appropriate to do so, but it's neither idiomatic nor necessary to put Get into the getter's name. If you have a field called owner (lower case, unexported), the getter method should be called Owner (upper case, exported), not GetOwner. The use of upper-case names for export provides the hook to discriminate the field from the method. A setter function, if needed, will likely be called SetOwner. Both names read well in practice

And technically this method is a getter. But I'm also fine with adding `map` prefix (I do prefer prefix myself since this method is actually a bit more than a simple getter).

zdaratom-s1 replied:
Yes I agree this is not getter, i takes argument so lets stick with map

tomaz-s1 replied:
Done.

zdaratom-s1 commented on exporter/datasetexporter/logs_exporter.go:
```diff
@@ -65,6 +66,73 @@ func buildBody(attrs map[string]interface{}, value pcommon.Value) string {
+func otelSeverityToDataSetSeverity(log plog.LogRecord) int {
+	// If log record doesn't contain severity or we can't map it to a valid DataSet severity,
+	// we use this value (INFO) instead
+	defaultSeverityLevel := 3
```
this constant can be extracted outside of the method.

tomaz-s1 replied:
Agreed, will move it to the file / module level.

tomaz-s1 replied:
Done.

zdaratom-s1 commented on exporter/datasetexporter/logs_exporter.go:
```diff
@@ -65,6 +66,73 @@ func buildBody(attrs map[string]interface{}, value pcommon.Value) string {
+	if sevNum > 0 {
+		// See https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitynumber
+		// for OTEL mappings
+		switch sevNum {
+		case 1, 2, 3, 4:
+			// TRACE
+			dataSetSeverity = 1
+		case 5, 6, 7, 8:
+			// DEBUG
+			dataSetSeverity = 2
+		case 9, 10, 11, 12:
+			// INFO
+			dataSetSeverity = 3
+		case 13, 14, 15, 16:
+			// WARN
+			dataSetSeverity = 4
+		case 17, 18, 19, 20:
+			// ERROR
+			dataSetSeverity = 5
+		case 21, 22, 23, 24:
+			// FATAL / CRITICAL / EMERGENCY
+			dataSetSeverity = 6
```
lets extract this block to separate method

tomaz-s1 replied:
Done.

zdaratom-s1 commented on exporter/datasetexporter/logs_exporter.go:
```diff
@@ -65,6 +66,73 @@ func buildBody(attrs map[string]interface{}, value pcommon.Value) string {
+	} else if sevText != "" {
+		// Per docs, SeverityNumber is optional so if it's not present we fall back to SeverityText
+		// https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext
+		switch strings.ToLower(sevText) {
+		case "fine", "finest":
+			dataSetSeverity = 0
+		case "trace":
+			dataSetSeverity = 1
+		case "debug":
+			dataSetSeverity = 2
+		case "info", "information":
+			dataSetSeverity = 3
+		case "warn", "warning":
+			dataSetSeverity = 4
+		case "error":
+			dataSetSeverity = 5
+		case "fatal", "critical", "emergency":
+			dataSetSeverity = 6
```
lets extract this block to separate method

tomaz-s1 replied:
Done.

zdaratom-s1 commented on exporter/datasetexporter/logs_exporter_test.go:
```diff
@@ -407,3 +403,155 @@ func TestConsumeLogsShouldSucceed(t *testing.T) {
 	)
 }
+
+func TestOtelSeverityToDataSetSeverityWithSeverityNumberNoSeverityText(t *testing.T) {
```
Consider parametrized test approach to improve readability, eg. https://elliotchance.medium.com/go-data-driven-or-parameterized-tests-cd2b62b21c8a

tomaz-s1 replied:
Yeah, I was a thinking about it, but I was a bit torn.

In tests, a lot of times, I personally try to avoid loops and parameterized approach since I've seen many issues over the years where there was a bug in the loop / parameters and the tests didn't actually end up executing all the intended paths / parameters. And code coverage wouldn't really bubble this up either (unless explicitly looking for it).

So while non-parameterized / non-loop approach does indeed mean more duplication, it's usually easier to spot issue like that.

zdaratom-s1 replied:
I am not that experienced with GO, but my goal is to improve readability of test, see eg java and junit5 aproach `@ValueSource and @ParametrizedTests` eg https://www.baeldung.com/parameterized-tests-junit-5#first-impression this provides very nice readability without any loops.

We might at least split the test so its bit readable eg. ```TestMappingOtelSeverityNumberAndNoTextToDataSetDebugSeverit, TestMappingOtelSeverityNumberAndNoTextToDataSetTraceSeverity and TestMappingInvalidOtelSeverityNumberAndNoTextToDataSetDefaultSeverity```

tomaz-s1 replied:
Done.

zdaratom-s1 commented on exporter/datasetexporter/logs_exporter.go:
```diff
@@ -65,6 +66,73 @@ func buildBody(attrs map[string]interface{}, value pcommon.Value) string {
+		// https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext
+		switch strings.ToLower(sevText) {
+		case "fine", "finest":
+			dataSetSeverity = 0
```
Consider defining DataSet severities as constants for better readability. It also may be reused in tests.

tomaz-s1 committed:
Update metadata and indicate plugin is distributed as part of otel
contrib.

tomaz-s1 committed:
Move default severity to be a module level constant instead of function
level constant.

tomaz-s1 committed:
Refactor / split otelSeverityToDataSetSeverity to two functions.

tomaz-s1 committed:
Also add test cases for invalid values.

tomaz-s1 committed:
Make it a module level constant.

tomaz-s1 committed:
Make dataset log levels constants for easier readability and maintaince
and update affected code and tests.

tomaz-s1 commented:
@zdaratom I pushed some commits which I believe should address all the feedback we agreed on (I didn't push map change yet + possible parameterized tests change since I had a bit different opinion on those and we didn't agree on the approach yet).

As I said before, I still plan to push one other changes in the near future (better / correct attribute name and value handling - aka getting rid non DataSet ideomatic approach of parameter.name and parameter.value).

tomaz-s1 committed:
Re-generate readme.

tomaz-s1 commented:
As discussed on Slack, I will work on that additional change (correct attribute field name and value handling in a separate PR in the very near future.

This way I can try and submit this PR upstream for review today and get CLA, etc. sorted out as soon as possible.

zdaratom-s1 reviewed (changes requested):
Please provide also changelog, see example https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23640/files#diff-09e45fe4784007dc8f80bb8585acec6638363b8b72bed0cca6d24876cb1e9a6c

tomaz-s1 committed:
Per PR review, add "map" suffix to the function names.

tomaz-s1 committed:
For readability, split different test scenarios in different test
functions. Also move functionality for making a test record into a
utility function.

tomaz-s1 commented:
@zdaratom "map" prefix should be addressed in 211db7e388fb3e45cf66a9fb3179d606b1cc3788 and test functions refactor / split in 656cc5fccdbc92d74bcaad1b4c9bb7e387ac11b9.

zdaratom-s1 commented:
Please provide also changelog, see example https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23640/files#diff-09e45fe4784007dc8f80bb8585acec6638363b8b72bed0cca6d24876cb1e9a6c without that PR to public repo wont be accepted

tomaz-s1 commented:
@zdaratom Thanks.

Yeah, I wanted to add changelog entry, but I saw it needs to reference github issue / pr and that won't be available until I open upstream PR. So I plan to do that once upstream PR is opened.

zdaratom-s1 commented:
I have referenced previously created issue [20660] so far it was enough

tomaz-s1 committed:
Add changelog entry file.
